### PR TITLE
Bump `ark-bn254` crate version from 0.4.0 to 0.5.0

### PR DESCRIPTION
**Pull request by samkim-crypto**
#### Problem
The crate `ark-bn254` version 0.5.0 has been published, which results in a number of performance improvements.

Version 0.4.0
```
bn128 addition          time:   [88.998 µs 89.050 µs 89.097 µs]
                        change: [+4381.5% +4391.4% +4401.4%] (p = 0.00 < 0.05)
                        Performance has regressed.
Found 17 outliers among 100 measurements (17.00%)
  4 (4.00%) low severe
  9 (9.00%) low mild
  1 (1.00%) high mild
  3 (3.00%) high severe

bn128 multiplication    time:   [113.95 µs 114.21 µs 114.46 µs]
                        change: [+63.427% +63.982% +64.553%] (p = 0.00 < 0.05)
                        Performance has regressed.
Found 19 outliers among 100 measurements (19.00%)
  8 (8.00%) low severe
  5 (5.00%) low mild
  1 (1.00%) high mild
  5 (5.00%) high severe

bn128 pairing           time:   [482.89 µs 483.06 µs 483.23 µs]
                        change: [+22.861% +24.481% +25.673%] (p = 0.00 < 0.05)
                        Performance has regressed.
Found 14 outliers among 100 measurements (14.00%)
  1 (1.00%) low severe
  5 (5.00%) low mild
  3 (3.00%) high mild
  5 (5.00%) high severe
```

Version 0.5.0
```
bn128 addition          time:   [1.9794 µs 1.9824 µs 1.9863 µs]
Found 12 outliers among 100 measurements (12.00%)
  5 (5.00%) low severe
  3 (3.00%) low mild
  4 (4.00%) high severe

bn128 multiplication    time:   [69.649 µs 69.840 µs 70.027 µs]
Found 26 outliers among 100 measurements (26.00%)
  4 (4.00%) low severe
  6 (6.00%) low mild
  2 (2.00%) high mild
  14 (14.00%) high severe

bn128 pairing           time:   [387.63 µs 393.63 µs 402.66 µs]
Found 12 outliers among 100 measurements (12.00%)
  3 (3.00%) high mild
  9 (9.00%) high severe
```

I particular, the bn128 addition operation improves by 45x.

#### Summary of Changes
Upgraded the `ark-*` packages to 0.5.0. I also upgraded `light-poseidon` to 0.3.0, which depends on the new `ark-*` 0.5.0 packages.

Fixes #

### CODE REVIEW

**Comment by t-nelson** [comment]
i don't think we can simply bump this version. if they introduced any divergence in calculations, consensus breaks. probably needs a feature gate, but i'm not too sure how we'd want to go about a simd for the change since implementation details are not part of the protocol

**Comment by samkim-crypto** [comment]
I used [solfuzz-agave](https://github.com/firedancer-io/solfuzz-agave) to check if the version bump introduces any breaking changes or diverging behavior.

```
# clone patched agave version for fuzzing
git clone https://github.com/firedancer-io/solfuzz-agave
git clone -b agave-v3.0.3-patches https://github.com/firedancer-io/agave

# bump the crate versions
cd agave
git apply << 'EOF'
diff --git a/Cargo.toml b/Cargo.toml
index 561eb5e4f0..4ef1b642e7 100644
--- a/Cargo.toml
+++ b/Cargo.toml
@@ -192,10 +192,10 @@ anyhow = "1.0.99"
 aquamarine = "0.6.0"
 arbitrary = "1.4.1"
 arc-swap = "1.7.1"
-ark-bn254 = "0.4.0"
-ark-ec = "0.4.0"
-ark-ff = "0.4.0"
-ark-serialize = "0.4.0"
+ark-bn254 = "0.5.0"
+ark-ec = "0.5.0"
+ark-ff = "0.5.0"
+ark-serialize = "0.5.0"
 array-bytes = "=1.4.1"
 arrayref = "0.3.9"
 arrayvec = "0.7.6"
@@ -301,7 +301,7 @@ libsecp256k1 = { version = "0.6.0", default-features = false, features = [
     "std",
     "static-context",
 ] }
-light-poseidon = "0.2.0"
+light-poseidon = "0.3.0"
 log = "0.4.27"
 lru = "0.7.7"
 lz4 = "1.28.1"
EOF

# update lock file
./scripts/cargo-for-all-lock-files.sh tree

# run fuzzer
cd ../solfuzz-agave
./scripts/generate_cargo.py --agave-path ../agave -o Cargo.toml
make binaries
./scripts/run_test_vectors.sh
```

The result shows that all test vectors pass:

```
❯ ./scripts/run_test_vectors.sh
++ nproc
+ NUM_PROCESSES=16
+ '[' '' == '' ']'
++ mktemp -d
+ LOG_PATH=/tmp/tmp.BmcaBL55ZO
+ echo 'LOG_PATH: /tmp/tmp.BmcaBL55ZO'
LOG_PATH: /tmp/tmp.BmcaBL55ZO
+ mkdir -p dump
+ '[' '!' -d dump/test-vectors ']'
+ cd dump/test-vectors
+ git pull -q
+ cd ../..
+ find dump/test-vectors/instr/fixtures -type f -name '*.fix'
+ xargs -P 16 -n 1000 ./target/release/test_exec_instr
+ find dump/test-vectors/txn/fixtures/precompile -type f -name '*.fix'
+ xargs -P 16 -n 1000 ./target/release/test_exec_txn
+ find dump/test-vectors/txn/fixtures/programs -type f -name '*.fix'
+ xargs -P 16 -n 1000 ./target/release/test_exec_txn
+ find dump/test-vectors/block/fixtures -type f -name '*.fix'
+ xargs -P 16 -n 1000 ./target/release/test_exec_block
+ find dump/test-vectors/syscall/fixtures -type f -name '*.fix'
+ xargs -P 16 -n 1000 ./target/release/test_exec_vm_syscall
+ find dump/test-vectors/vm_interp/fixtures -type f -name '*.fix'
+ xargs -P 16 -n 1000 ./target/release/test_exec_vm_interp
+ find dump/test-vectors/elf_loader/fixtures -type f -name '*.fix'
+ xargs -P 16 -n 1000 ./target/release/test_exec_elf_loader
++ grep -wR FAIL /tmp/tmp.BmcaBL55ZO
++ wc -l
+ failed=0
++ grep -wR OK /tmp/tmp.BmcaBL55ZO
++ wc -l
+ passed=140901
+ echo Test vectors success
Test vectors success
```

@t-nelson The ark-bn254 version update doesn't seem to add any breaking changes and the solfuzz-agave test vectors all pass as well. Do you think we can get this PR merged? We need this PR to address [simd-359](https://github.com/solana-foundation/solana-improvement-documents/pull/367).

cc @vadorovsky

**Comment by vadorovsky** [comment]
Not clicking "approve" (I'll let Trent decide), but I'm in favor of this.

We could also consider vendoring/forking github.com/arkworks/algebra, like we did with quinn, so we could review the changes incoming in next versions. What do you think?

**Comment by t-nelson** [comment]
are we confident that we in fact have adequate coverage here? there are [176 commits](https://github.com/arkworks-rs/algebra/compare/v0.4.0...v0.5.0) between these two versions. can we get fd opinions as to whether a feature gate would be acceptable here?

**Comment by samkim-crypto** [comment]
@0x0ece Do you think we can upgrade these crates behind a feature gate?

If we were to do a feature gate, then it might make sense to do it as part of the https://github.com/solana-foundation/solana-improvement-documents/pull/367 implementation.

**Comment by 0x0ece** [comment]
I agree with Sam that a gate seems unnecessary if it passes test-vectors (it's more than 3k tests covering a lot of annoying edge cases).

**Comment by samkim-crypto** [comment]
Since we need to update behavior of the poseidon syscall behind a feature gate anyway, I think it wouldn't hurt to just do the crate upgrade behind the same feature gate. @vadorovsky created https://github.com/anza-xyz/agave/pull/8534, so if any of you guys have some time to look at that PR, then it would be great.

I'll close this PR once https://github.com/anza-xyz/agave/pull/8534 lands.

**Comment by samkim-crypto** [comment]
Closing in favor of https://github.com/anza-xyz/agave/pull/8534.
